scripts/prepare-restricted-environment.sh

```diff
@@ -0,0 +1,132 @@
+my_registry="$HOST"
+
+# if able to run sudo and update your ca trust store
+# oc get secret -n openshift-ingress  router-certs-default -o go-template='{{index .data "tls.crt"}}' | base64 -d | sudo tee /etc/pki/ca-trust/source/anchors/${HOST}.crt  > /dev/null
+# sudo update-ca-trust enable
+# sudo podman login -u kubeadmin -p "$(oc whoami -t)" "$HOST"
+
+# else just login with tls disabled
+podman login -u kubeadmin -p "$(oc whoami -t)" --tls-verify=false "$HOST"
+
+# Create local directory
+mkdir -p "${my_catalog}/${prod_operator_package_name}"
+
+echo "[INFO] Fetching metadata for the ${prod_operator_package_name} operator catalog channel, packages, and bundles."
+opm render "${prod_operator_index}" \
+  | jq "select \
+    (\
+      (.schema == \"olm.bundle\" and .name == \"${prod_operator_bundle_name}.${prod_operator_version}\") or \
+      (.schema == \"olm.package\" and .name == \"${prod_operator_package_name}\") or \
+      (.schema == \"olm.channel\" and .package == \"${prod_operator_package_name}\") \
+    )" \
+  | jq "select \
+     (.schema == \"olm.channel\" and .package == \"${prod_operator_package_name}\").entries \
+      |= [{name: \"${prod_operator_bundle_name}.${prod_operator_version}\"}]" \
+  > "${my_catalog}/${prod_operator_package_name}/render.json"
+
+echo "[DEBUG] Got $(cat "${my_catalog}/${prod_operator_package_name}/render.json" | wc -l) lines of JSON from the index!"
+# echo "[DEBUG] Got this from the index:
+# ======"
+# cat "${my_catalog}/${prod_operator_package_name}/render.json"
+# echo "======"
+
+echo "[INFO] Creating the catalog dockerfile."
+if [ -f "${my_catalog}.Dockerfile" ]; then
+  rm -f "${my_catalog}.Dockerfile"
+fi
+opm generate dockerfile "./${my_catalog}"
+
+echo "[INFO] Building the catalog image locally."
+podman build -t "${my_operator_index}" -f "./${my_catalog}.Dockerfile" --no-cache .
+
+echo "[INFO] Disabling the default Red Hat Ecosystem Catalog."
+oc patch OperatorHub cluster --type json \
+    --patch '[{"op": "add", "path": "/spec/disableAllDefaultSources", "value": true}]'
+
+echo "[INFO] Deploying your catalog image to the $my_operator_index registry."
+# See: https://docs.openshift.com/container-platform/latest/installing/disconnected_install/installing-mirroring-installation-images.html#olm-mirroring-catalog_installing-mirroring-installation-images
+### TODO fix this step or switch to oc adm catalog mirror?
+# https://docs.openshift.com/container-platform/4.14/installing/disconnected_install/installing-mirroring-installation-images.html#installation-images-samples-disconnected-mirroring-assist_installing-mirroring-installation-images ?
+# FATA[0010] writing manifest: uploading manifest v1.1.0 to default-route-openshift-image-registry.apps.ci-ln-x0yk982-72292.origin-ci-int-gce.dev.rhcloud.com/rhdh/rhdh-index: denied 
+# skopeo copy --src-tls-verify=false --dest-tls-verify=false --all "containers-storage:$my_operator_index" "docker://$my_operator_index"
```

Comment:
I got the same "denied" error, and after looking into the OCP Image Registry logs, I found that it was expecting a namespace `rhdh` to be available, in order to be able to push an image named `default-route-openshift-image-registry.apps.ci-ln-x0yk982-72292.origin-ci-int-gce.dev.rhcloud.com/rhdh/rhdh-index`.
I had to do the same thing for all the images for we are calling skopeo:
```
for ns in rhdh rhel9 openshift4 rh-osbs; do oc create namespace $ns; done
```

But even after that, it made some progress, but then I hit another issue:
```
FATA[0015] copying image 1/4 from manifest list: writing signatures: downloading signatures for sha256:5ef1720092d3b8431c7eee5a5ffb336a9a3d6329cc17ee4aaa1a2f18b42267ef in default-route-opensh
ift-image-registry.apps.7737beef7661f70d0e22.hypershift.aws-2.ci.openshift.org/rhel9/postgresql-15: StatusCode: 404, ""
```
Then I tried to work around this error by adding `--remove-signatures` to skopeo, but I hit another error for some of the images:
```
"application/vnd.oci.image.index.v1+json" to be written to destination, but we cannot modify it: "Instructed to preserve digests"
```

So to me, it seems we cannot use the internal OpenShift image registry as a target registry, as mentioned in the docs: https://docs.openshift.com/container-platform/4.14/installing/disconnected_install/installing-mirroring-installation-images.html#installation-about-mirror-registry_installing-mirroring-installation-images

---

After taking a closer look at the [`prepare-restricted-environment.sh` script from DevSpaces
](https://access.redhat.com/webassets/avalon/d/Red_Hat_OpenShift_Dev_Spaces-3.10-Administration_guide-en-US/files/prepare-restricted-environment.sh), it seems to me that it requires a mirror registry as a prerequisite.

I managed to get this script to work after:
- deploying a registry in the same cluster (but on another namespace)
- commenting out all the commands in this script where we were relying on the OpenShift internal registry

```
$ ./scripts/prepare-restricted-environment.sh --prod_operator_index "quay.io/rhdh/iib:latest-v4.14-x86_64" --prod_operator_package_name "rhdh" --prod_operator_bundle_name "rhdh-operator" --prod_operator_version "v1.1.0" --my_registry "airgap-registry-airgap-helper-ns.apps.ci-ln-s0zlv6t-72292.origin-ci-int-gce.dev.rhcloud.com"

[...]
[INFO] Creating CatalogSource and ImageContentSourcePolicy
catalogsource.operators.coreos.com/rhdh-disconnected-install created
imagecontentsourcepolicy.operator.openshift.io/rhdh-disconnected-install created
[INFO] Catalog airgap-registry-airgap-helper-ns.apps.ci-ln-s0zlv6t-72292.origin-ci-int-gce.dev.rhcloud.com/rhdh/rhdh-index:v1.1.0 deployed to the airgap-registry-airgap-helper-ns.apps.ci-ln-s0zlv6t-72292.origin-ci-int-gce.dev.rhcloud.com registry.
```

So I think we could maybe change this script so that it requires a registry to be passed. And create a separate helper script that would deploy a registry in the cluster, for folks that do not already have a mirror registry.
@nickboldt WDYT ?

Reply:
> deploy a registry in the cluster

yeah, that would be great as a default behaviour (and then optionally allow users to bring their own registry)

Reply:
That makes sense - I'll update it accordingly.